aten/src/ATen/functorch/BatchRulesReduceOps.cpp

```diff
@@ -368,58 +421,80 @@ std::tuple<Tensor,optional<int64_t>> searchsorted_batch_rule(
+//   if exists. Otherwise, the value is ignored.
+#define REDUCTION_BOXED_ARGS(op, dim_pos, keepdim_case, maybe_keepdim_pos) \
+  m.impl(#op, torch::CppFunction::makeFromBoxedFunction< \
+      SINGLE_ARG(boxed_reduction_batch_rule<dim_pos, keepdim_case, maybe_keepdim_pos>)>());
+
+// Provided for your convenience; most operators that have a keepdim arg
+// will work with this macro.
+// Assumes the dim arg is at position 1 and the keepdim arg is at pos 2.
+#define REDUCTION_WITH_KEEPDIM_ARG(op) \
+  REDUCTION_BOXED_ARGS(op, 1, KEEPDIM_CASE_VARIABLE, 2)
+
+// Provided for your convenience; most operators that do not have a keepdim
+// arg will work with this macro.
+// Assumes the dim arg is at position 1 and the operation always returns
+// a tensor of the same rank (instead of a smaller rank).
+#define REDUCTION_NO_KEEPDIM_ARG(op) \
+  REDUCTION_BOXED_ARGS(op, 1, KEEPDIM_CASE_TRUE, -1)
+
 TORCH_LIBRARY_IMPL(aten, FuncTorchBatched, m) {
   VMAP_SUPPORT2(searchsorted, Tensor, searchsorted_batch_rule);
-  REDUCTION_BOXED(_fft_r2c);
-  REDUCTION_BOXED(_fft_c2r);
-  REDUCTION_BOXED(_fft_c2c);
-  REDUCTION_BOXED(amax);
-  // REDUCTION_BOXED(aminmax); Currently fails due to inconsistent scalar semantics.
-  REDUCTION_BOXED(amin);
-  REDUCTION_BOXED(any.dim);
-  REDUCTION_BOXED(argmax);
-  REDUCTION_BOXED(argmin);
-  REDUCTION_BOXED(count_nonzero.dim_IntList);
-  REDUCTION_BOXED(cummax);
-  REDUCTION_BOXED(cummin);
-  REDUCTION_BOXED(cumprod);
-  REDUCTION_BOXED(cumsum);
+  REDUCTION_NO_KEEPDIM_ARG(_fft_r2c);
+  REDUCTION_NO_KEEPDIM_ARG(_fft_c2r);
+  REDUCTION_NO_KEEPDIM_ARG(_fft_c2c);
+  REDUCTION_WITH_KEEPDIM_ARG(amax);
+  REDUCTION_WITH_KEEPDIM_ARG(amin);
+  REDUCTION_WITH_KEEPDIM_ARG(any.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(argmax);
+  REDUCTION_WITH_KEEPDIM_ARG(argmin);
+  REDUCTION_BOXED_ARGS(count_nonzero.dim_IntList, 1, KEEPDIM_CASE_FALSE, -1);
+  REDUCTION_NO_KEEPDIM_ARG(cummax);
+  REDUCTION_NO_KEEPDIM_ARG(cummin);
+  REDUCTION_NO_KEEPDIM_ARG(cumprod);
+  REDUCTION_NO_KEEPDIM_ARG(cumsum);
   m.impl("dist", dist_decomp);
-  REDUCTION_BOXED_ARGS(kthvalue, 2);
-  REDUCTION_BOXED_ARGS(linalg_vector_norm, 2);
-  REDUCTION_BOXED(log_softmax.int);
-  REDUCTION_BOXED(logcumsumexp);
-  REDUCTION_BOXED(logsumexp);
+  REDUCTION_BOXED_ARGS(kthvalue, 2, KEEPDIM_CASE_VARIABLE, 3);
+  REDUCTION_BOXED_ARGS(linalg_vector_norm, 2, KEEPDIM_CASE_VARIABLE, 3);
+  REDUCTION_NO_KEEPDIM_ARG(log_softmax.int);
+  REDUCTION_NO_KEEPDIM_ARG(logcumsumexp);
+  REDUCTION_WITH_KEEPDIM_ARG(logsumexp);
   m.impl("max", max_decomp);
-  REDUCTION_BOXED(max.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(max.dim);
   m.impl("mean", mean_decomp);
-  REDUCTION_BOXED(mean.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(mean.dim);
   m.impl("median", median_decomp);
-  REDUCTION_BOXED(median.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(median.dim);
   m.impl("min", min_decomp);
-  REDUCTION_BOXED(min.dim);
-  REDUCTION_BOXED(mode);
+  REDUCTION_WITH_KEEPDIM_ARG(min.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(mode);
```

Comment:
(nothing needed for this PR) does this suggest that the original tests would have missed these as well? It might be good to have a follow up to extend the op infos for ones we need as a followup

Reply:
The original tests missed all of these cases. The edge case is vmap over a dimension of shape 1. The problem is not that the OpInfos are missing the test cases (they have them), the problem is that vmap testing generates dimensions of size 2 instead of size 1.

I didn't want to increase the testing time by adding vmap over dimension of shape 1 testing to all OpInfos. While we do not directly test this case, it is indirectly tested because the boxed_reduction_batch_rule is now written in a bdim-size-agnostic way.